### ISSUE TITLE
Errors caused by X when subsetting AnnData object

### ISSUE THREAD

**Issue by jonas2612** [scverse/anndata#1140]
### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [ ] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

I haven't been able to reproduce the error(s) on a smaller example. The dataset can be downloaded and assembled from here https://shendure-web.gs.washington.edu/content/members/cxqiu/public/nobackup/jax/download/adata/. `X` is in csr und `experimental_batch` is saved as categories. I receive two different error messages dependent on subsetting by equality or inequality

Code:

```python
import scanpy as sc

adata = sc.read('/lustre/groups/ml01/workspace/monge_velo/data/mouse_gastrulation_atlas.h5ad')
adata_tmp1 = adata[adata.obs["experimental_batch"]!='run_22']
adata_tmp1.X
```

Traceback:

```pytb
File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/anndata/_core/anndata.py:612, in AnnData.X(self)
    609     X = None
    610 elif self.is_view:
    611     X = as_view(
--> 612         _subset(self._adata_ref.X, (self._oidx, self._vidx)),
    613         ElementRef(self, "X"),
    614     )
    615 else:
    616     X = self._X

File ~/mambaforge/envs/atlas_pca/lib/python3.10/functools.py:889, in singledispatch.<locals>.wrapper(*args, **kw)
    885 if not args:
    886     raise TypeError(f'{funcname} requires at least '
    887                     '1 positional argument')
--> 889 return dispatch(args[0].__class__)(*args, **kw)

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/anndata/_core/index.py:140, in _subset_spmatrix(a, subset_idx)
    138 if len(subset_idx) > 1 and all(isinstance(x, cabc.Iterable) for x in subset_idx):
    139     subset_idx = (subset_idx[0].reshape(-1, 1), *subset_idx[1:])
--> 140 return a[subset_idx]

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_index.py:75, in IndexMixin.__getitem__(self, key)
     73         return self._get_arrayXint(row, col)
     74     elif isinstance(col, slice):
---> 75         return self._get_arrayXslice(row, col)
     76 else:  # row.ndim == 2
     77     if isinstance(col, INT_TYPES):

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_csr.py:335, in _csr_base._get_arrayXslice(self, row, col)
    333     col = np.arange(*col.indices(self.shape[1]))
    334     return self._get_arrayXarray(row, col)
--> 335 return self._major_index_fancy(row)._get_submatrix(minor=col)

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_compressed.py:705, in _cs_matrix._major_index_fancy(self, idx)
    702 np.cumsum(row_nnz, out=res_indptr[1:])
    704 nnz = res_indptr[-1]
--> 705 res_indices = np.empty(nnz, dtype=idx_dtype)
    706 res_data = np.empty(nnz, dtype=self.dtype)
    707 csr_row_index(M, indices, self.indptr, self.indices, self.data,
    708               res_indices, res_data)

ValueError: negative dimensions are not allowed
```

Code:

```python
import scanpy as sc

adata = sc.read('/lustre/groups/ml01/workspace/monge_velo/data/mouse_gastrulation_atlas.h5ad')
adata_tmp2 = adata[adata.obs["experimental_batch"]=='run_22']
adata_tmp2.X
```

Traceback:

```pytb
File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/anndata/_core/anndata.py:612, in AnnData.X(self)
    609     X = None
    610 elif self.is_view:
    611     X = as_view(
--> 612         _subset(self._adata_ref.X, (self._oidx, self._vidx)),
    613         ElementRef(self, "X"),
    614     )
    615 else:
    616     X = self._X

File ~/mambaforge/envs/atlas_pca/lib/python3.10/functools.py:889, in singledispatch.<locals>.wrapper(*args, **kw)
    885 if not args:
    886     raise TypeError(f'{funcname} requires at least '
    887                     '1 positional argument')
--> 889 return dispatch(args[0].__class__)(*args, **kw)

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/anndata/_core/index.py:140, in _subset_spmatrix(a, subset_idx)
    138 if len(subset_idx) > 1 and all(isinstance(x, cabc.Iterable) for x in subset_idx):
    139     subset_idx = (subset_idx[0].reshape(-1, 1), *subset_idx[1:])
--> 140 return a[subset_idx]

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_index.py:75, in IndexMixin.__getitem__(self, key)
     73         return self._get_arrayXint(row, col)
     74     elif isinstance(col, slice):
---> 75         return self._get_arrayXslice(row, col)
     76 else:  # row.ndim == 2
     77     if isinstance(col, INT_TYPES):

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_csr.py:335, in _csr_base._get_arrayXslice(self, row, col)
    333     col = np.arange(*col.indices(self.shape[1]))
    334     return self._get_arrayXarray(row, col)
--> 335 return self._major_index_fancy(row)._get_submatrix(minor=col)

File ~/mambaforge/envs/atlas_pca/lib/python3.10/site-packages/scipy/sparse/_compressed.py:707, in _cs_matrix._major_index_fancy(self, idx)
    705 res_indices = np.empty(nnz, dtype=idx_dtype)
    706 res_data = np.empty(nnz, dtype=self.dtype)
--> 707 csr_row_index(M, indices, self.indptr, self.indices, self.data,
    708               res_indices, res_data)
    710 return self.__class__((res_data, res_indices, res_indptr),
    711                       shape=new_shape, copy=False)

ValueError: Output dtype not compatible with inputs.
```

### Versions

<details>

```
-----
anndata             0.9.2
numpy               1.24.4
pandas              2.0.3
scanpy              1.9.3
scipy               1.11.2
session_info        1.0.0
-----
PIL                         10.0.0
asciitree                   NA
asttokens                   NA
backcall                    0.2.0
cloudpickle                 2.2.1
comm                        0.1.2
cycler                      0.10.0
cython_runtime              NA
cytoolz                     0.12.2
dask                        2023.8.1
dateutil                    2.8.2
debugpy                     1.5.1
decorator                   5.1.1
entrypoints                 0.4
executing                   0.8.3
fasteners                   0.18
h5py                        3.9.0
importlib_metadata          NA
ipykernel                   6.19.2
jedi                        0.18.1
jinja2                      3.1.2
joblib                      1.3.2
kiwisolver                  1.4.4
llvmlite                    0.40.1
lz4                         4.3.2
markupsafe                  2.1.3
matplotlib                  3.7.2
mpl_toolkits                NA
msgpack                     1.0.5
natsort                     8.4.0
numba                       0.57.1
numcodecs                   0.11.0
packaging                   23.1
parso                       0.8.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
platformdirs                3.10.0
prompt_toolkit              3.0.36
psutil                      5.9.5
ptyprocess                  0.7.0
pure_eval                   0.2.2
pyarrow                     12.0.1
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.6.0
pydevd_concurrency_analyser NA
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.11.2
pyparsing                   3.0.9
pytz                        2023.3
six                         1.16.0
sklearn                     1.3.0
stack_data                  0.2.0
tblib                       1.7.0
threadpoolctl               3.2.0
tlz                         0.12.2
toolz                       0.12.0
tornado                     6.3.3
traitlets                   5.7.1
typing_extensions           NA
wcwidth                     0.2.5
yaml                        6.0
zarr                        2.16.1
zipp                        NA
zmq                         23.2.0
zoneinfo                    NA
-----
IPython             8.8.0
jupyter_client      7.4.8
jupyter_core        5.1.1
-----
Python 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
Linux-4.18.0-425.3.1.el8.x86_64-x86_64-with-glibc2.28
```

</details>

**Comment by ivirshup** [comment]
Interesting issue! You've caught most of us traveling, so no chance to replicate yet. I suspect this may have to do with some recent internal changes in scipy I made... I'll be interested to see if pinning scipy lower helps fix this.

As there is no `mouse_gastrulation_atlas.h5ad` file in the folder you link, what does that correspond to? Is it the each of those files concatenated together?

**Comment by jonas2612** [comment]
Yes, exactly. The file is the concatenation of all JAX-files

**Comment by ivirshup** [comment]
Does this happen with any one of the files, or do they have to be concatenated for this to occur?

**Comment by jonas2612** [comment]
To merge the files I first concatenated two files each with `anndata`, but the memory requirements for concatenating the results are too large, so I used a function of my own to concatenate `.X`. I currently believe that there the issue originates from, as I can subset the intermediate files. The function I used was

```python
from pathlib import Path

import h5py
from scipy import sparse

import anndata as ad
from anndata._core.sparse_dataset import SparseDataset
from anndata.experimental import read_elem, write_elem


def read_everything_but_X(pth) -> ad.AnnData:
    attrs = ["obs", "var", "obsm", "varm", "obsp", "varp", "uns"]
    with h5py.File(pth) as f:
        adata = ad.AnnData(**{k: read_elem(f[k]) for k in attrs})
    return adata


def concat_on_disk(input_pths: list[Path], output_pth: Path):
    """
    Params
    ------
    input_pths
        Paths to h5ad files which will be concatenated
    output_pth
        File to write as a result
    """
    annotations = ad.concat([read_everything_but_X(pth) for pth in input_pths])

    annotations.write_h5ad(output_pth)
    n_variables = annotations.shape[1]
    
    del annotations

    with h5py.File(output_pth, "a") as target:
        dummy_X = sparse.csr_matrix((0, n_variables), dtype="float64")
        dummy_X.indptr = dummy_X.indptr.astype("int64") # Guarding against overflow for very large datasets

        write_elem(target, "X", dummy_X)
        mtx = SparseDataset(target["X"])
        for p in pths:
            with h5py.File(p, "r") as src:
                mtx.append(SparseDataset(src["X"]))
``` 
Alas, I need `.X`, so the lazy concatenation from `anndata` isn't an option.

**Comment by ivirshup** [comment]
I've gotten a chance to look at this. But haven't had a chance to use the exact same file.

* First, this is an 11 million cell dataset, so it has pretty large memory requirements to load and index the whole thing. I haven't managed to create a session with enough memory to that yet.
* The exact file this was encountered with may no longer exist. @jonas2612 has recreated it with `concat_on_disk` however.

With the new file, I am able to do things like:

```python
adata = ad.read_h5ad(ADATA_PTH, backed="r")
result = adata[adata.obs["experimental_batch"]!='run_22'].to_memory()
```

Or:

```python
half = adata[:5_000_000].to_memory()
not_run_19 = half[half.obs["experimental_batch"] != "run_19"].copy()
```

So maybe this works with new file. Or maybe I'm not hitting the right conditions yet. This vaguely looks like a recent issue in scipy where index dtypes weren't large enough, which I'd like to know we're avoiding https://github.com/scipy/scipy/issues/19205.

@jonas2612 could you try and replicate the original issue with the new file before I try and dig in some more?

**Comment by colinshew** [comment]
Hi, I think I am also encountering this issue. I am getting the exact same traceback as the first one in the original post.

My dataset is large but not so much as 11M cells. Sorry this is not easily reproducible; the anndata was processed through a method for aligning datasets across species (SAMap), which is why there are so many genes (duplicates are needed to consider pairs of genes). However, maybe this extra information is useful:

```
# adata is an anndata of dimension 116312 × 29452
# g2 is a list of genes with duplicates occurring in adata.var_names of length 490563
adata[:, g2]             # allowed, but can't access matrix X:
adata[:, g2].X           # *** ValueError: negative dimensions are not allowed

# subset by gene
adata[:, g2[:372585]].X  # ok (len=372585)
adata[:, g2[:372586]].X  # *** ValueError: negative dimensions are not allowed
adata[:, g2[113515:]].X  # ok (len=377048)
adata[:, g2[113514:]].X  # *** ValueError: negative dimensions are not allowed

# subset by cell
adata[:50000, g2].X       # ok
adata[50000:, g2].X       # ok
```

I.e. subsets of the data are fine, but too large and I get the negative dimension error when trying to access `.X`. I also have anndata objects processed in parallel with fewer cells/genes that do not have this error. Does this seem like an overflow issue then? I also need the full matrix -- any thoughts would be appreciated!

```
-----
anndata             0.9.2
pandas              2.1.1
samalg              1.0.0
samap               1.0.15
session_info        1.0.0
-----
PIL                         10.0.1
asttokens                   NA
backcall                    0.2.0
cffi                        1.16.0
colorama                    0.4.6
comm                        0.1.2
cycler                      0.12.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.5.1
decorator                   5.1.1
defusedxml                  0.7.1
entrypoints                 0.4
executing                   0.8.3
fast_histogram              0.11
h5py                        3.8.0
harmonypy                   NA
hnswlib                     NA
igraph                      0.10.8
importlib_resources         NA
ipykernel                   6.19.2
ipython_genutils            0.2.0
ipywidgets                  8.0.4
jedi                        0.18.1
joblib                      1.3.2
kiwisolver                  1.4.5
leidenalg                   0.10.1
llvmlite                    0.39.1
matplotlib                  3.8.0
mpl_toolkits                NA
natsort                     8.4.0
networkx                    3.1
numba                       0.56.3
numpy                       1.23.5
packaging                   22.0
parso                       0.8.3
patsy                       0.5.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
platformdirs                2.5.2
plotly                      5.17.0
prompt_toolkit              3.0.36
psutil                      5.9.0
ptyprocess                  0.7.0
pure_eval                   0.2.2
pycparser                   2.21
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.6.0
pydevd_concurrency_analyser NA
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.11.2
pynndescent                 0.5.10
pyparsing                   3.1.1
pytz                        2023.3.post1
scanpy                      1.9.3
scipy                       1.11.2
six                         1.16.0
sklearn                     1.3.0
stack_data                  0.2.0
statsmodels                 0.14.0
texttable                   1.6.7
threadpoolctl               3.2.0
tornado                     6.2
tqdm                        4.66.1
traitlets                   5.7.1
typing_extensions           NA
umap                        0.5.3
wcwidth                     0.2.5
yaml                        6.0.1
zipp                        NA
zmq                         23.2.0
zoneinfo                    NA
-----
IPython             8.8.0
jupyter_client      7.4.8
jupyter_core        5.1.1
notebook            6.5.6
-----
Python 3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:49:32) [GCC 12.3.0]
Linux-3.10.0-1160.95.1.el7.x86_64-x86_64-with-glibc2.17
```

**Comment by jonas2612** [comment]
Yes, I will try to recreate the file tomorrow. I'll let you know once it's build

**Comment by dnjst** [comment]
Hi, I can say I have been having this issue as well.

It happens when my anndata's X slot is in CSR format. If I do adata.X = adata.X.toarray(), the subsetting works, but of course this blows up my memory usage so I cannot do this.

if I do adata.X = csc_matrix(adata.X), it doesn't seem to fix the issue...

Any suggestions on what versions of scipy or anndata might help with this?

**Comment by ilan-gold** [comment]
@dnjst @jonas2612 We need a reproducer for this, it seems both data and code.  However, if `concat_on_disk` can solve your problem, or `concat` + `dask`, I think we could consider this closed until someone comes along with something more concrete.

**Comment by jonas2612** [comment]
I remember solving this with a couple of line of code later. I don't have access to the environment anymore, but if I can recreate my approach I'll send in the code

**Comment by jonas2612** [comment]
@dnjst Can you try if subsetting your anndata object before and then loading the subsetted anndata object to memory works, so

`adata_in_mem = adata[obs_indices].to_memory()`

This worked for me in the end but is very slow though, as it is randomly accessing data instead of reading them sequentially

**Comment by quentinblampey** [comment]
Hi, I had the same error, but from a slightly different use case.

I have a small AnnData object (`55917 × 18651`), but I try to access many indices. More precisely, I wanted to run `adata[:, genes].X[indices]`, where `indices` has 2M indices. (NB: running `adata[indices, genes].X` made my kernel crash, but I guess it's expected).

After some investigation, I found out the issue may be related to the dtype of the indices underlying the sparse matrix X, potentially causing an integer overflow. More precisely, I used int64 instead of int32:
```python
X.indices = X.indices.astype(np.int64)
X.indptr = X.indptr.astype(np.int64)
```
Then, performing my index selection worked nicely!
In my specific case, it seems that the issue is not related to AnnData, but scipy
